### DISCUSSION TITLE
Release / Prerelease action triggers

### DISCUSSION THREAD

**Discussion by wstyres** [community/community#26281] [Actions]
<p>I’m having some trouble writing triggers for when I make a new release of my project. I have two workflows that are triggered on different releases (Releases vs. Pre-Releases). I only want the Pre-Release Workflow to run on Releases that are marked as a pre-release and only run the Release Workflow on releases not marked as a pre-release.</p>
<p>I’ve read the documentation and saw that Actions supports this functionality but I’ve set it up and the Release workflow runs on both Pre-Release and actual releases. I’ve linked both below for reference, any ideas?</p>
<p>Release: <a href="https://github.com/wstyres/Zebra/blob/master/.github/workflows/release.yml#L5" rel="nofollow noopener">https://github.com/wstyres/Zebra/blob/master/.github/workflows/release.yml#L5</a></p>
<p>Pre-Release: <a href="https://github.com/wstyres/Zebra/blob/master/.github/workflows/beta-release.yml#L5" rel="nofollow noopener">https://github.com/wstyres/Zebra/blob/master/.github/workflows/beta-release.yml#L5</a></p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>There is not a <strong>released</strong> type for release event, if use <strong>published</strong> instead,  when create a pre-released version, we also need to publish it. So <strong>published</strong> type workflow will be trigged both Pre-Release and actual releases.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/78f75d4c66893864e1aa9d53f3ffedeb5537a89b.png" data-download-href="/uploads/short-url/hg7gSRDi9fJOQDkA7yWMKtheEuT.png?dl=1" title="pre-release.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181096635-2754b9a0-b762-455f-ade2-cecaba509cd0.png" alt="pre-release.png" data-base62-sha1="hg7gSRDi9fJOQDkA7yWMKtheEuT" width="520" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/7/78f75d4c66893864e1aa9d53f3ffedeb5537a89b_2_520x500.png, https://aws1.discourse-cdn.com/github/original/2X/7/78f75d4c66893864e1aa9d53f3ffedeb5537a89b.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/7/78f75d4c66893864e1aa9d53f3ffedeb5537a89b.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/78f75d4c66893864e1aa9d53f3ffedeb5537a89b_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">pre-release.png</span><span class="informations">689×662 22.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>As a work around, in published type workflow,  you could add if condition to filter pre-release attribute. If it equals false , then run the job. If is true, then skip the job.</p>
<pre><code>if: "!github.event.release.prerelease"
</code></pre>
<p>For example:</p>
<p><img src="https://user-images.githubusercontent.com/24806478/181096637-2cbdf68f-6b29-4229-b58b-a4035030589d.png" alt="work around.png" data-base62-sha1="6UqvD8eZ8VKmaydAAbmumjElHCc" width="406" height="388"></p>
<p>I need to mention that this work around could not stop workflow to be triggered, just stop the job to run. When a pre-release published, both two workflows are triggered, but the job in Upload Release will be skipped and the whole workflow run shows failed.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/b/b1c47be4ef1573fc2e1a3491c8c368f7bb00f21d.png" data-download-href="/uploads/short-url/pmBtAq4VVCe6XXAV1Spjv8YHlSd.png?dl=1" title="work around limit.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181096638-7ad3d2bb-436a-4cd0-b498-58e2b514e10c.png" alt="work around limit.png" data-base62-sha1="pmBtAq4VVCe6XXAV1Spjv8YHlSd" width="690" height="95" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/b/b1c47be4ef1573fc2e1a3491c8c368f7bb00f21d_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">work around limit.png</span><span class="informations">845×117 8.32 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Reply by Andre601** [reply]
Hey, thanks for the share of this.

Tho, I myself currently encounter a really strange situation where the output of `github.event.prerelease` is the oposite of what it should be?

Like I have this test-setup here:
```yaml
name: Test Release action

on:
  release:
    types: [published]

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
    - name: Runs on Pre-Release
      if: 'github.event.prerelease'
      run: echo "This is a Pre-Release"
    - name: Runs on Release
      if: '!github.event.prerelease'
      run: echo "This is a Release"
```
When I now publish a pre-release is the Step "Runs on Release" triggered while the other step is skipped... So it's basically the oposite of what it should be...

Edit: I do see my error here. Forgot the `release` part.

**Comment by wstyres** [comment]
<p>Or is there a way to have one workflow run and use that if tag to run one or the other? The only real change between the two workflows is where I’m CDing into a directory (beta vs. repo) So if there were a simplier way to combine both that would work as well</p>

**Comment by CKGrafico** [comment]
<p>Is not a really good solution but thank you! I hope github will allow us to have a prereleased and released types in the near future :slight_smile:</p>

**Comment by Yanjingzhu** [comment]
<p><a class="mention" href="/ckgrafico">@ckgrafico</a> Good news! GitHub engineering team is working on adding released type to Release event. I would be on production in the near future. Although I don’t have a timeline for it , you could keep an eye on <a href="https://github.blog/changelog/" rel="nofollow noopener">https://github.blog/changelog/</a> for new released features. </p>

**Comment by ianaz** [comment]
<p>Thanks, this will do for us.</p>
<p>For the “pre-release” -&gt; “release” (by unchecking the box and releasing after a pre-release) we had to listen to <code>edited</code> event too. The problem there is that edited doesn’t take into consideration the checkbox. If we edit the text, uncheck and save, then it will trigger <code>edited</code>, but if only the checkbox is unchecked then not</p>

**Comment by matthewfeickert** [comment]
<p>Follow up on this: If I have a workflow (c.f. <a href="https://github.com/scikit-hep/pyhf/issues/1907" class="inline-onebox" rel="noopener nofollow ugc">Add pre-release support to publish workflow · Issue #1907 · scikit-hep/pyhf · GitHub</a>) that I want to run on</p>
<pre><code class="lang-yaml">if: github.event_name == 'release' &amp;&amp; github.event.action == 'published'
</code></pre>
<p>this will run on normal GitHub releases, but does not run on GitHub pre-releases. Does this mean that as there <em>is</em> the <code>boolean</code> <code> github.event.release.prerelease</code> that things that if <code> github.event.release.prerelease</code> is <code>true</code> then <code>github.event.action</code> will <em>not</em> be <code>published</code>?</p>

**Comment by victorlin** [comment]
5 years later, there are now `prereleased` and `released` event types:

![image](https://github.com/community/community/assets/13424970/c70e0351-4ea9-4787-9ed9-56389d7cc119)

![release](https://github.com/community/community/assets/13424970/6d51ae35-e1f7-45ed-bb15-0439823f9450)

References:

- [release - Events that trigger workflows - GitHub Docs](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#release)
- [release - Webhook events and payloads - GitHub Docs](https://docs.github.com/en/webhooks/webhook-events-and-payloads?actionType=released#release)
